extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java

```diff
@@ -0,0 +1,169 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.auth.quickconnect.utility;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.guacamole.GuacamoleClientException;
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+
+public class QCParser {
+
+    /**
+     * The default protocol to parse to if one is undefined.
+     */
+    public static final String DEFAULT_URI_PROTOCOL = "ssh";
+
+    /**
+     * The default host to use if one is not defined.
+     */
+    public static final String DEFAULT_URI_HOST = "localhost";
+
+    /**
+     * The default port to use if one is not defined.
+     */
+    public static final Integer DEFAULT_URI_PORT = 22;
+
+    /**
+     * Parse out a URI string and get a connection from that string,
+     * or an exception if the parsing fails.
+     *
+     * @param uri
+     *     The string form of the URI to be parsed.
+     *
+     * @return
+     *     A GuacamoleConfiguration using a combination of the parsed
+     *     URI values and default values when not specified in the
+     *     URI.
+     *
+     * @throws GuacamoleException
+     *     When an error occurs parsing the URI.
+     */
+    public static GuacamoleConfiguration getConfiguration(String uri)
+            throws GuacamoleException {
+
+        URI qcUri;
+        try {
+            qcUri = new URI(uri);
+        }
+        catch (URISyntaxException e) {
+            throw new GuacamoleClientException("Invalid URI Syntax", e);
+        }
+        String protocol = qcUri.getScheme();
+        String host = qcUri.getHost();
+        Integer port = qcUri.getPort();
+        String userInfo = qcUri.getUserInfo();
+        String query = qcUri.getQuery();
+        String username = null;
+        String password = null;
+        List<String> paramList = null;
+
+        if (protocol == null || protocol.equals(""))
+            protocol = DEFAULT_URI_PROTOCOL;
+
+        if (host == null || host.equals(""))
+            host = DEFAULT_URI_HOST;
+
+        if (port == -1 || port < 1)
+            port = DEFAULT_URI_PORT;
+
+        if (query != null && !query.equals(""))
+            paramList = Arrays.asList(query.split("&"));
+
+        if (userInfo != null && !userInfo.equals("")) {
+            String[] authenticators = userInfo.split(":");
+            if (authenticators.length > 0 && authenticators[0] != null)
+                username = authenticators[0];
+            if (authenticators.length > 1 && authenticators[1] != null)
```

Comment:
What if `authenticators` is something like "myusername:myPassword:That:Has:Colons"? Rather than manually parse everything with `split()`, I suggest defining (and pre-compiling) regular expressions. It would avoid recompiling the pattern each time `split()` is invoked, and may be more robust.

Alternatively, though I haven't looked into this, there may be a better class for parsing URIs which handles all of this. It doesn't look like anything we're doing here is non-standard with respect to URI format.

Reply:
I haven't found a class that parses the username and password into separate components - even the URI and URL classes that have methods for grabbing the "user-info" part of the string (including both username and password) don't seem to have methods for splitting it out to individual components.  I'll work on the regex route.

Reply:
Okay, implemented regex and Pattern/Matcher for this.